packages/react/src/components/Tree/Tree.tsx

```diff
@@ -0,0 +1,331 @@
+    }
+  }
+
+  getInitialAutoControlledState() {
+    return { activeItemIds: [] }
+  }
+
+  treeRef = React.createRef<HTMLElement>()
+
+  handleTreeItemOverrides = (predefinedProps: TreeItemProps) => ({
+    onTitleClick: (e: React.SyntheticEvent, treeItemProps: TreeItemProps) => {
+      let { activeItemIds } = this.state
+      const { id, siblings } = treeItemProps
+      const { exclusive } = this.props
+
+      if (!Tree.hasSubtree(treeItemProps)) {
+        return
+      }
+
+      const activeItemIdIndex = activeItemIds.indexOf(id)
+
+      if (activeItemIdIndex > -1) {
+        activeItemIds = [
+          ...activeItemIds.slice(0, activeItemIdIndex),
+          ...activeItemIds.slice(activeItemIdIndex + 1),
+        ]
+      } else {
+        if (exclusive) {
+          siblings.some(sibling => {
+            const activeSiblingIdIndex = activeItemIds.indexOf(sibling['id'])
+            if (activeSiblingIdIndex > -1) {
+              activeItemIds = [
+                ...activeItemIds.slice(0, activeSiblingIdIndex),
+                ...activeItemIds.slice(activeSiblingIdIndex + 1),
+              ]
+              return true
+            }
+            return false
+          })
+        }
+
+        activeItemIds = [...activeItemIds, id]
+      }
+
+      this.setState({
+        activeItemIds,
+      })
+
+      _.invoke(predefinedProps, 'onTitleClick', e, treeItemProps)
+    },
+    onFocusParent: (e: React.SyntheticEvent, treeItemProps: TreeItemProps) => {
+      const { parent } = treeItemProps
+
+      if (!parent) {
+        return
+      }
+
+      const { itemsForRender } = this.state
+      const elementToBeFocused = itemsForRender[parent['id']].elementRef
+
+      if (!elementToBeFocused) {
+        return
+      }
+
+      elementToBeFocused.current.focus()
+      _.invoke(predefinedProps, 'onFocusParent', e, treeItemProps)
+    },
+    onFocusFirstChild: (e: React.SyntheticEvent, treeItemProps: TreeItemProps) => {
+      const { id } = treeItemProps
+
+      const { itemsForRender } = this.state
+      const currentElement = itemsForRender[id].elementRef
+
+      if (!currentElement && currentElement.current) {
+        return
+      }
+
+      const elementToBeFocused = getNextElement(this.treeRef.current, currentElement.current)
+
+      if (!elementToBeFocused) {
+        return
+      }
+
+      elementToBeFocused.focus()
+      _.invoke(predefinedProps, 'onFocusFirstChild', e, treeItemProps)
+    },
+    onSiblingsExpand: (e: React.SyntheticEvent, treeItemProps: TreeItemProps) => {
+      if (this.props.exclusive) {
+        return
+      }
+
+      const { id, siblings } = treeItemProps
+      const { activeItemIds } = this.state
+
+      siblings.forEach(sibling => {
+        if (Tree.hasSubtree(sibling) && activeItemIds.indexOf(sibling['id']) < 0) {
+          activeItemIds.push(sibling['id'])
+        }
+      })
+
+      if (Tree.hasSubtree(treeItemProps) && activeItemIds.indexOf(id) < 0) {
```

Comment:
nit: `activeItemIds.indexOf` have been used 5 times to either check if item is active or not. How about to export this to func?
```
isActiveItem = (id): boolean => {
   const { activeItemIds } = this.state
    return activeItemIds.indexOf(id) > -1
}
```